### ISSUE TITLE
Adaptive Batch Size Schedules for Distributed Training of Language Models with Data and Model Parallelism

### ISSUE THREAD

**Issue by Hiroki11x** [Hiroki11x/Papers#402]
## Links
https://openreview.net/forum?id=pi2TiX7er9#discussion

## Authors
[Tim Tsz-Kit Lau](https://arxiv.org/search/cs?searchtype=author&query=Lau,+T+T), [Weijian Li](https://arxiv.org/search/cs?searchtype=author&query=Li,+W), [Chenwei Xu](https://arxiv.org/search/cs?searchtype=author&query=Xu,+C), [Han Liu](https://arxiv.org/search/cs?searchtype=author&query=Liu,+H), [Mladen Kolar](https://arxiv.org/search/cs?searchtype=author&query=Kolar,+M)

## TLDR
- 本研究は、大規模言語モデルの事前学習における 適応的バッチサイズスケジュール (adaptive batch size schedules) を提案する。
- PyTorch FSDPを用いた実装により、数十億パラメータ規模のモデルでも効率的かつ汎用的に利用可能であることを示す。
- 実験により、固定バッチやヒューリスティックなウォームアップよりも優れた汎化性能と効率を達成し、Adamに対する理論的収束保証も提示する。

### 背景と課題
- 大規模言語モデル (LLM) の学習では 大規模バッチ が計算効率を高める一方で、汎化性能が低下する「generalization gap」が問題となる（図1参照, p.2）。
- 既存手法は「固定大バッチ」または「ヒューリスティックなウォームアップ」に依存しており、学習動態に適応できない。

### 提案手法の新規性

- 適応的バッチサイズスケジュール (Adaptive Batch Size Schedules) を理論的根拠に基づいて設計。
- データ並列 (DDP) だけでなく、モデル並列 (FSDP) にも適用可能な「2次元並列 (2D parallelism)」を実現。
- Adam最適化に対する理論的収束保証を提示（従来はSGDやAdaGradに限定）。

#### （Adaptive Batch Size Schedules）の決定方法

- 本研究では、Gradient Noise Scale (GNS) のように「ノイズスケールを推定して臨界バッチサイズを決める」わけではありません。
- 代わりに、Norm Test (分散ノルム条件) に基づいて「バッチ勾配が十分に真の勾配方向を近似できているか」を逐次判定します。

$$
\frac{1}{b} \sum_{i \in B} \left\| \nabla \ell_i(w) - \nabla \mathcal{L}_B(w) \right\|^2 
\leq \eta^{2} \left\| \nabla \mathcal{L}_B(w) \right\|^2
$$

#### Notation
- $b$: 現在のバッチサイズ  
- $B$: バッチに含まれるサンプルのインデックス集合  
- $\ell_i(w)$: サンプル $i$ に対する損失関数  
- $\nabla \ell_i(w)$: サンプル $i$ に対する勾配  
- $\{L}_B(w) = \frac{1}{b}\sum_{i \in B}\ell_i(w)$ : バッチ $B$ における平均損失  
- $\nabla \mathcal{L}_B(w)$: バッチ勾配 (平均勾配)  
- $\eta \in [0,1)$: 許容する勾配方向誤差のパラメータ

- が成り立たなければ、次ステップのバッチサイズを増加させます。
- この条件は「勾配近似の分散が小さく、 descent direction を保てるか」を直接テストするもので、逐次的にバッチサイズを拡大する仕組みです。
- DDPやFSDPを用いる場合は、各ワーカーのミニバッチ勾配の分散を all-reduce で集約して推定する（DDP-Norm, FSDP-Norm）形に拡張されています。

### 実証方法 (Contribution の検証)

- PyTorch FSDP上に効率的実装 (DDP-Norm, FSDP-Norm) を構築。
- MicroLlama-300M, TinyLlama-1.1B, OpenLlama-3Bを対象にC4データセットで学習実験。

### 結果:

- 提案法は 固定バッチ (4096や8192) より低い検証損失を達成。
- ウォームアップ (2.5–2.5–95%) と同等かそれ以上の性能を示すが、ヒューリスティック設計不要で自動適応。
- 大規模モデル (TinyLlama 1.1B, OpenLlama 3B) でも 損失スパイク抑制 と汎化性能改善を確認。

### 図1 (p.2): 
バッチサイズが大きいほど汎化性能が悪化する「generalization gap」を視覚化。

<img width="557" height="402" alt="Image" src="https://github.com/user-attachments/assets/7c43c1c3-cb19-4e84-b1de-fe273b79b959" />

### 図2 (p.8, MicroLlama 300M): 
提案法では初期は小バッチ、後半は大バッチに自動移行し、効率と汎化性能を両立。

<img width="749" height="293" alt="Image" src="https://github.com/user-attachments/assets/0a8aa8f8-b200-43ca-94d9-e10a2cf421af" />

### 図3 (p.9, TinyLlama 1.1B): 
提案法はバッチ8192使用時の悪化を防ぎ、4096に近い性能を維持。

<img width="743" height="305" alt="Image" src="https://github.com/user-attachments/assets/ea74a325-6620-46b9-980b-fb87abb48d0f" />

### 図4 (p.9, OpenLlama 3B): 
提案法は固定8192よりも低い検証損失を達成しつつ、トレーニング効率も確保。

<img width="734" height="301" alt="Image" src="https://github.com/user-attachments/assets/8a98e8d7-fead-46dc-a990-27df16dc695e" />

## 結論

- 提案手法は、汎用的・理論的に正当化されたバッチサイズスケジューリングを提供し、LLMの大規模事前学習を効率化。
- 将来的にはPyTorch FSDP2やDTensorとの統合で、さらに大規模 (>7B パラメータ) モデルへの拡張が可能。
